modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/StaticUnitDeployer.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.ignite.internal.logger.Loggers;
+
+/**
+ * Observes a predefined directory with statically provisioned deployment units and
+ * registers their presence in the deployment store.
+ *
+ * <p>The observer scans the {@code deploymentUnitsRoot} directory, expecting the following structure:
+ * {@code <unitId>/<version>/}
+ * For every discovered {@code unitId}-{@code version} pair that is not yet registered for this node,
+ * the observer creates (or reuses) a cluster status and then creates the node status marked as
+ * {@link DeploymentStatus#DEPLOYED}.
+ */
+public class StaticUnitDeployer {
+    private static final IgniteLogger LOG = Loggers.forClass(StaticUnitDeployer.class);
+
+    private final DeploymentUnitStore deploymentUnitStore;
+
+    private final String nodeName;
+
+    private final Path deploymentUnitsRoot;
+
+    /**
+     * Constructor.
+     */
+    public StaticUnitDeployer(
+            DeploymentUnitStore deploymentUnitStore,
+            String nodeName,
+            Path deploymentUnitsRoot
+    ) {
+        this.deploymentUnitStore = deploymentUnitStore;
+        this.nodeName = nodeName;
+        this.deploymentUnitsRoot = deploymentUnitsRoot;
+    }
+
+    /**
+     * Scans the filesystem for statically deployed units and registers their cluster and node statuses
+     * if they are not yet present in the store.
+     *
+     * <p>Already registered unit versions for this node are skipped. New ones are registered as DEPLOYED.
+     */
+    public CompletableFuture<Void> searchAndDeployStaticUnits() {
+        LOG.info("Start search static deployment units.");
+        StaticUnits allUnits = collectStaticUnits();
+
+        return deploymentUnitStore.getNodeStatuses(nodeName).thenCompose(statuses -> {
+            List<CompletableFuture<?>> futures = new ArrayList<>();
+
+            for (UnitNodeStatus status : statuses) {
+                allUnits.filter(status.id(), status.version());
+            }
+
+            allUnits.forEach((id, version) -> {
+                LOG.info("Start processing unit {}:{}", id, version);
+                CompletableFuture<Boolean> future = deploymentUnitStore.createClusterStatus(id, version, Set.of(nodeName))
+                        .thenCompose(status -> {
+                            if (status == null) {
+                                return deploymentUnitStore.getClusterStatus(id, version).thenCompose(it ->
+                                        deploymentUnitStore.createNodeStatus(nodeName, id, version, it.opId(), DEPLOYED)
+                                );
+                            } else {
+                                return deploymentUnitStore.createNodeStatus(nodeName, id, version, status.opId(), DEPLOYED);
+                            }
+                        })
+                        .whenComplete((result, t) ->
+                                LOG.info("Finished static status creating {}:{} with result {}", t, id, version, result)
```

Comment:
"Static status" sounds confusing.

I think it can be useful to log a full list of detected static deployment units in the very end.

One message per unit seems excessive.